### ISSUE TITLE
JsonException when parsing single value JSON

### ISSUE THREAD

**Issue by acheaito** [rest-assured/rest-assured#949]
According to [RFC 7159](https://tools.ietf.org/html/rfc7159#section-2), "A JSON text is a sequence of tokens.  The set of tokens includes six structural characters, strings, numbers, and three literal names."
As such, TRUE, FALSE, NULL, Strings, and numbers should be acceptable JSON even if the response does not start with { or [. For example, a RESTFul response of **false** is valid JSON. 

However, RestAssured fails when parsing such single token JSON responses. Below is the exception trace:
```
groovy.json.JsonException: A JSON payload should start with an openning curly brace '{' or an openning square bracket '['.
Instead, 'false' was found on line: 1, column: 1
       at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
       at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
       at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
       at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
       at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:83)
       at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:105)
       at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:247)
       at io.restassured.internal.path.json.ConfigurableJsonSlurper.parse(ConfigurableJsonSlurper.groovy:114)
       at io.restassured.internal.path.json.ConfigurableJsonSlurper$parse.callCurrent(Unknown Source)
       at io.restassured.internal.path.json.ConfigurableJsonSlurper.parseText(ConfigurableJsonSlurper.groovy:83)
       at 
```

**Comment by johanhaleby** [comment]
REST Assured is using an old json parser (JsonSlurper) from Groovy. I tried the new one and then it works but then I get a lot of other tests that fails. Need to look into it more. Thanks for pointing out, I actually didn't know this.

**Comment by nikolmarku** [comment]
Hi, How to setup new parser? Any progress on this ?

**Comment by johanhaleby** [comment]
Unfortunately not. It's much harder to use the new JsonSlurper since (from what I remember when I tried to implement this) it doesn't provide hooks to modify it's behavior (which we need to do in order to convert BigDecimals to floats/doubles). The only option that I can think of right now is to essentially copy most of the Groovy code into REST Assured. And this is not something that I'm particularly keen on doing.

**Comment by rfftrifork** [comment]
Any news on this issue? I would consider it a rather essential bug if the framework doesn't parse json correctly, which is basically the case. Would it by any chance be possible to get a workaround at least?

**Comment by johanhaleby** [comment]
@rfftrifork Unfortunately still no updates. You could use any json parser that you like though by either by using [object mapping](https://github.com/rest-assured/rest-assured/wiki/Usage#object-mapping) or manually:

```java
InputStream is = when().get("/json").asInputStream();
MyJsonLibrary myJsonLib = MyJsonLibrary(is); 
...
```

**Comment by shan-github-project** [comment]
`
io.restassured.path.json.exception.JsonPathException: Failed to parse the JSON document
	at io.restassured.path.json.JsonPath$ExceptionCatcher.invoke(JsonPath.java:1002)
	at io.restassured.path.json.JsonPath$4.doParseWith(JsonPath.java:967)
	at io.restassured.path.json.JsonPath$JsonParser.parseWith(JsonPath.java:1047)
	at io.restassured.path.json.JsonPath.get(JsonPath.java:202)
	at files.DynamicJson.addbook(DynamicJson.java:23)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:583)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:719)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:989)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:648)
	at org.testng.TestRunner.run(TestRunner.java:505)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:455)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:450)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:415)
	at org.testng.SuiteRunner.run(SuiteRunner.java:364)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:84)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1208)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1137)
	at org.testng.TestNG.runSuites(TestNG.java:1049)
	at org.testng.TestNG.run(TestNG.java:1017)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:114)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:251)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:77)
Caused by: groovy.json.JsonException: Lexing failed on line: 1, column: 1, while reading '<', no possible valid JSON value or punctuation could be recognized.
	at groovy.json.JsonLexer.nextToken(JsonLexer.java:86)
	at groovy.json.JsonLexer$nextToken.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:130)
	at io.restassured.internal.path.json.ConfigurableJsonSlurper.parse(ConfigurableJsonSlurper.groovy:97)
	at io.restassured.internal.path.json.ConfigurableJsonSlurper$parse.callCurrent(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
	at io.restassured.internal.path.json.ConfigurableJsonSlurper.parseText(ConfigurableJsonSlurper.groovy:83)
	at io.restassured.path.json.JsonPath$4$1.method(JsonPath.java:965)
	at io.restassured.path.json.JsonPath$ExceptionCatcher.invoke(JsonPath.java:1000)
	... 29 more
`

Any Update on this i am  facing issue of Parsing Json.

**Comment by MonaAgrawal** [comment]
Any update on this issue, I am facing the same issue.

**Comment by godoshian** [comment]
It's almost the five year anniversary of this issue! How should we celebrate? Lol jk sorry I'm a noob but seriously I am also facing this issue.

**Comment by cygri** [comment]
I'm facing this too. Quite surprised to see a JSON-centric test framework using a non-compliant JSON parser.

I've never used Groovy, but from reading the source, the problem might be fixable with a simple change in `ConfigurableJsonParser.groovy` around [here](https://github.com/rest-assured/rest-assured/blob/master/json-path/src/main/groovy/io/restassured/internal/path/json/ConfigurableJsonSlurper.groovy#L104). This reads:
```
    if (token.getType() == OPEN_CURLY) {
      content = parseObject(lexer)
    } else if (token.getType() == OPEN_BRACKET) {
      content = parseArray(lexer)
    } else {
      throw new JsonException(
              "A JSON payload should start with " + OPEN_CURLY.getLabel() +
                      " or " + OPEN_BRACKET.getLabel() + ".\n" +
                      "Instead, '" + token.getText() + "' was found " +
                      "on line: " + token.getStartLine() + ", " +
                      "column: " + token.getStartColumn()
      )
    }
```
Propose to add an additional condition before `} else {`:
```
} else if (token.getType().ordinal() >= NULL.ordinal()) {
  content = currentToken.getValue()
```
This is how primitive JSON values are handled further down in `parseObject` and `parseArray`, so it should work here too.

I'm not able to test this for lack of Groovy knowledge, but maybe someone else can?

**Comment by jekkel** [comment]
FYI There's a [related groovy bug](https://issues.apache.org/jira/browse/GROOVY-11123) which would hinder the proposed fix from working for plain numeric values.

**Comment by jekkel** [comment]
FYI The underlying groovy issue is fixed with Groovy 4.0.14. The linked PR updates groovy and is now a complete fix of the issue.
